### PR TITLE
PYIC-3207: Add CiScoring function to CRI return flow

### PR DESCRIPTION
**Pull request by lwebb-iig**
Adds new CiScoring function to the CriReturnStepFunction to perform CiScoring prior to evaluate gpg45

## Proposed changes

Run the new lambda as part of the CRI return flow, before the evaluate-gpg45-scores lambda itself. At this point it’s ok if the CI scoring is done twice, once by the new lambda and once by the evaluate-gpg45-scores lambda.

If a user has breaching CIs the new lambda should handle that. If the user doesn’t it should allow the flow to move to the evaluate-gpg45-scores.

### What changed

- Adds CiScoring function prior to evaluate gpg45
- Adds in relevant permissions and lambda env vars required to run

Example flow with no breaching CIs:
image

Example flow with breaching CIs:
image

### Why did it change

Only run evaluategpg45 scoring when required - exit journey early if breaching CIs

### Issue tracking

- [PYIC-3207](https://govukverify.atlassian.net/browse/PYIC-3207)

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

### Other considerations

- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

[PYIC-3207]: https://govukverify.atlassian.net/browse/PYIC-3207?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deploy/template.yaml [hunk 1]
```diff
@@ -2385,18 +2422,76 @@ Resources:
...
       Policies:
         - VPCAccessPolicy: { }
+        - KMSDecryptPolicy:
+            KeyId: !Ref DynamoDBKmsKey
+        - DynamoDBReadPolicy:
+            TableName: !Ref SessionsTable
+        - DynamoDBWritePolicy:
+            TableName: !Ref SessionsTable
+        - DynamoDBReadPolicy:
+            TableName: !Ref UserIssuedCredentialsV2Table
```
Comment (rpayal) [inline deploy/template.yaml]: Do we need this one ?

File: lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java [hunk 2]
```diff
@@ -202,7 +202,7 @@ void shouldReturnJourneyResponseOnSuccessfulRequest() throws Exception {
         verify(verifiableCredentialJwtValidator)
                 .validate(any(SignedJWT.class), eq(testPassportIssuer), eq(TEST_USER_ID));
 
-        assertEquals("/journey/evaluate", output.get("journey"));
+        assertEquals("/journey/ci-scoring", output.get("journey"));
```
Comment (rpayal) [inline lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java]: Can use already declared JourneyUris for constant.

File: lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java [hunk 3]
```diff
@@ -632,7 +632,7 @@ void shouldPassNullApiKeyWhenCriDoesNotRequireApiKey() throws Exception {
                         testCriNotRequiringApiKey,
                         null,
                         CREDENTIAL_ISSUER_ID);
-        assertEquals("/journey/evaluate", output.get("journey"));
+        assertEquals("/journey/ci-scoring", output.get("journey"));
```
Comment (rpayal) [inline lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java]: Can use already declared JourneyUris for constant.

File: lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java [hunk 4]
```diff
@@ -658,7 +658,7 @@ void shouldReturnJourneyEvaluateResponseOnSuccessfulPendingCriResponse() {
 
         Map<String, Object> output = handler.handleRequest(testInput, context);
 
-        assertEquals("/journey/evaluate", output.get("journey"));
+        assertEquals("/journey/ci-scoring", output.get("journey"));
```
Comment (rpayal) [inline lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java]: Same....as previous.

File: deploy/template.yaml (hunk 1 above)
Comment (lwebb-iig) [reply deploy/template.yaml]: Nope, will remove

File: deploy/template.yaml [hunk 5]
```diff
@@ -2385,18 +2422,76 @@ Resources:
           - !GetAtt LambdaSecurityGroup.GroupId
       Policies:
         - VPCAccessPolicy: { }
+        - KMSDecryptPolicy:
+            KeyId: !Ref DynamoDBKmsKey
+        - DynamoDBReadPolicy:
+            TableName: !Ref SessionsTable
+        - DynamoDBWritePolicy:
```
Comment (therealslimsladey) [inline deploy/template.yaml]: Don't think we're writing to sessions?

File: deploy/template.yaml [hunk 6]
```diff
@@ -2385,18 +2422,76 @@ Resources:
...
+            TableName: !Ref SessionsTable
+        - DynamoDBReadPolicy:
+            TableName: !Ref UserIssuedCredentialsV2Table
+        - DynamoDBReadPolicy:
+            TableName: !Ref ClientOAuthSessionsTable
+        - SSMParameterReadPolicy:
+            ParameterName: !Sub ${Environment}/core/*
+        - SQSSendMessagePolicy:
```
Comment (therealslimsladey) [inline deploy/template.yaml]: No audit messages being sent?

File: lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java (hunk 2 above)
Comment (therealslimsladey) [inline lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java]: Use a constant for the expected value? It's in several places.

**Review by therealslimsladey — CHANGES_REQUESTED** [review]
Equivalent change needed to wire in ProcessAsyncCredentialHandler to hand off to CiScoring (instead of EvaluateGpg45) as well or is that on a separate PR?

**Comment by lwebb-iig** [comment]
> Equivalent change needed to wire in ProcessAsyncCredentialHandler to hand off to CiScoring (instead of EvaluateGpg45) as well or is that on a separate PR?

@therealslimsladey Separate PR, this is purely focused on integrating with the CriReturnStepFunction of which the ProcessAsyncCredentialHandler is not part of

**Review by rpayal — APPROVED** [review]
LGTM

**Review by therealslimsladey — APPROVED** [review]
lgtm

**Pull request merged**